File: 150.Evaluate Reverse Polish Notation.java
/**
 *Basic idea:
 * stack
 *Result:
 * 20 / 20 test cases passed.
 * Status: Accepted
 * Runtime: 13 ms
 * Your runtime beats 94.33% of javasubmissions.
 *Date:
 * 9/8/2016
 */

public class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        for(String token : tokens){
            switch(token){
                case "+":
                    stack.push(stack.pop()+stack.pop());
                    break;
                case "-":
                    stack.push(-stack.pop()+stack.pop());
                    break;
                case "*":
                    stack.push(stack.pop()*stack.pop());
                    break;
                case "/":
                    Integer a = stack.pop();
                    Integer b = stack.pop();
                    stack.push(b/a);
                    break;
                default:
                    stack.push(Integer.parseInt(token));
            }
        }
        return stack.pop();
    }
}

File: 150.Evaluate Reverse Polish Notation.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []
        op = {
            '+': int.__add__,
            '-': int.__sub__,
            '*': int.__mul__,
            '/': lambda x, y: int(float(x) / y)
        }
        for token in tokens:
            if token.isdigit() or len(token) > 1:
                stack.append(int(token))
            else:
                v2 = stack.pop()
                v1 = stack.pop()
                stack.append(op[token](v1, v2))
        return stack[0]
